File: android/app/src/main/java/com/magicforms/mobile/data/FormDraftStore.kt
package com.magicforms.mobile.data

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class FormDraft(
    val textValues: Map<String, String> = emptyMap(),
    val boolValues: Map<String, Boolean> = emptyMap(),
    val checklistValues: Map<String, List<String>> = emptyMap(),
)

object FormDraftStore {
    private val json = Json { ignoreUnknownKeys = true }
    private const val PREFS = "form_drafts"

    fun storageKey(formId: Int?): String? = formId?.let { "form_$it" }

    fun storageKey(relatedAccessToken: String?): String? =
        relatedAccessToken?.takeIf { it.isNotBlank() }?.let { "related_$it" }

    fun load(context: Context, key: String?): FormDraft? {
        if (key.isNullOrBlank()) return null
        val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(key, null)
            ?: return null
        return runCatching { json.decodeFromString<FormDraft>(raw) }.getOrNull()
    }

    fun save(
        context: Context,
        key: String?,
        textValues: Map<String, String>,
        boolValues: Map<String, Boolean>,
        checklistValues: Map<String, Set<String>>,
    ) {
        if (key.isNullOrBlank()) return
        val draft = FormDraft(
            textValues = textValues,
            boolValues = boolValues,
            checklistValues = checklistValues.mapValues { (_, v) -> v.sorted() },
        )
        val encoded = json.encodeToString(draft)
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(key, encoded)
            .apply()
    }

    fun clear(context: Context, key: String?) {
        if (key.isNullOrBlank()) return
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().remove(key).apply()
    }
}

File: ios/MagicFormsMobile/Services/FormDraftStore.swift
import Foundation

struct FormDraft: Codable {
    var textValues: [String: String] = [:]
    var boolValues: [String: Bool] = [:]
    var checklistValues: [String: [String]] = [:]
    var savedAt: Date = Date()
}

enum FormDraftStore {
    private static let prefix = "form_draft_"

    static func key(formId: Int?) -> String? {
        guard let formId else { return nil }
        return "\(prefix)form_\(formId)"
    }

    static func key(relatedAccessToken: String?) -> String? {
        guard let relatedAccessToken, !relatedAccessToken.isEmpty else { return nil }
        return "\(prefix)related_\(relatedAccessToken)"
    }

    static func load(storageKey: String?) -> FormDraft? {
        guard let storageKey,
              let data = UserDefaults.standard.data(forKey: storageKey) else { return nil }
        return try? JSONDecoder().decode(FormDraft.self, from: data)
    }

    static func save(
        storageKey: String?,
        textValues: [String: String],
        boolValues: [String: Bool],
        checklistValues: [String: Set<String>]
    ) {
        guard let storageKey else { return }
        let draft = FormDraft(
            textValues: textValues,
            boolValues: boolValues,
            checklistValues: checklistValues.mapValues { Array($0).sorted() }
        )
        guard let data = try? JSONEncoder().encode(draft) else { return }
        UserDefaults.standard.set(data, forKey: storageKey)
    }

    static func clear(storageKey: String?) {
        guard let storageKey else { return }
        UserDefaults.standard.removeObject(forKey: storageKey)
    }
}
